concordium-contracts-common/src/schema_json.rs

```diff
@@ -1,47 +1,306 @@
+#[derive(thiserror::Error, Debug, Clone)]
+pub enum ToJsonError {
+    /// JSON formatter failed to represent value.
+    FormatError,
+    /// Failed to deserialize data to type expected from schema.
+    DeserialError {
+        position: u32,
+        schema:   Type,
+        reason:   String,
+        data:     ToJsonErrorData,
+    },
+    /// Trace leading to the original [ToJsonError].
+    TraceError {
+        position: u32,
+        schema:   Type,
+        error:    Box<ToJsonError>,
+    },
+}
+
+impl Display for ToJsonError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            ToJsonError::FormatError => write!(f, "Failed to format as JSON"),
+            ToJsonError::DeserialError {
+                position,
+                schema,
+                reason,
+                data,
+            } => write!(
+                f,
+                "Failed to deserialize {:?} due to: {} - from position {} of bytes {}",
+                schema, reason, position, data
+            ),
+            ToJsonError::TraceError {
+                ..
+            } => write!(f, "{}", self.print(false)),
+        }
+    }
+}
+
+impl TraceError for ToJsonError {
+    fn display_layer(&self, verbose: bool) -> (String, Option<&Self>) {
+        if let ToJsonError::TraceError {
+            error,
+            position,
+            schema,
+        } = self
+        {
+            let message = if verbose {
+                format!("In deserializing position {} into type {:?}", position, schema)
+            } else {
+                format!("{:?}", schema)
+            };
+
+            return (message, Some(error));
+        }
+
+        let message = format!("{}", self);
+        (message, None)
+    }
+
+    fn get_inner_error(&self) -> Option<&Self> {
+        if let ToJsonError::TraceError {
+            error,
+            ..
+        } = self
+        {
+            return Some(error);
+        }
+
+        None
+    }
+}
+
+impl ToJsonError {
+    /// Wraps a [`ToJsonError`] in a [`ToJsonError::TraceError`], providing a
+    /// trace to the origin of the error.
+    fn add_trace(self, position: u32, schema: &Type) -> Self {
+        ToJsonError::TraceError {
+            position,
+            schema: schema.clone(),
+            error: Box::new(self),
+        }
+    }
+
+    /// Returns a formatted error message for variant.
+    /// [`ToJsonError::TraceError`] supports printing a verbose form
+    /// including a more detailed description of the error stack, which is
+    /// returned if `verbose` is set to true.
+    pub fn print(&self, verbose: bool) -> String { self.display_nested(verbose) }
+
+    /// Gets the underlying error of a [`ToJsonError::TraceError`]. For any
+    /// other variant, this simply returns the error itself.
+    pub fn get_error(&self) -> &Self { self.get_innermost_error() }
+}
+
+/// Result with error type [`ToJsonError`]
```

Comment:
I think it would make sense to add an example or two of errors that can occcur. I think that would help explain the format pretty quickly. You can just take one of the nested examples from the test.

If we add that documentation here or to the method then it will be present in the rendered documentation on docs.rs which help make it understandable I think.

Reply:
Good idea

Reply:
Would it not make more sense to add this documentation to the actual error instead of the result, though?

Reply:
I added examples for both `ToJsonError` and `JsonError` in each of their documentation blocks.

Reply:
> Would it not make more sense to add this documentation to the actual error instead of the result, though?

Sure. Whatever is most discoverable in the rendered documentation is fine with me.

Reply:
Actually the reason I suggested to add it to the type was that it is a type alias.

So in the siganture of the function it will just show ToJsonResult<A> and the error part will not be clickable directly.

So I thought it would not be easily discoverable.

Reply:
Good point. I will remove the type alias instead.